mm/pmm.c

```diff
@@ -596,14 +573,69 @@ static void merge_frames(frame_t *first) {
     merge_frames(first);
 }
 
+static inline bool enough_4k_frames(void) {
+    frame_t *frame;
+    int count = 0;
+
+    list_for_each_entry (frame, &free_frames[PAGE_ORDER_4K], list) {
+        if (++count >= MIN_NUM_4K_FRAMES)
+            return true;
+    }
+
+    return false;
+}
+
+static void try_create_4k_frames(void) {
+    while (!enough_4k_frames()) {
+        frame_t *frame = find_larger_frame(free_frames, PAGE_ORDER_4K);
+        if (!frame)
+            panic("No more frames available to create 4K frames");
+        split_frame(frame);
+    }
+}
+
+/* Reserves and returns the first free frame fulfilling
+ * the condition specified by the callback.
+ * This function does not split larger frames.
+ */
+frame_t *get_free_frames_cond(free_frames_cond_t cb) {
+    spin_lock(&lock);
+    for_each_order (order) {
+        frame_t *frame;
+
+        if (list_is_empty(&free_frames[order])) {
+            BUG_ON(order == PAGE_ORDER_4K);
+            continue;
+        }
+
+        if (order == PAGE_ORDER_4K)
+            try_create_4k_frames();
```

Comment:
As `for_each_order()` starts at `PAGE_ORDER_4K`, you can simply move the call to `try_create_4k_frames()` out of the loop and call it unconditionally.

Reply:
Fixed. Thanks.